### PR TITLE
[Feature](bangc-ops): add ms_deform_attn_forward source code.

### PR DESCRIPTION
**Pull request by Unireverse**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Add ms_deform_attn_forward source code.

## 2. Modification

bangc-ops/kernels/ms_deform_attn_forward
bangc-ops/test/mluop_gtest/pb_gtest/src/zoo/ms_deform_attn_forward
bangc-ops/mlu_op.h

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

- [ float ] diff1: diff1 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results           |
|----------------|-----------------------------------|--------------------------------------|
|        1       |Supported hardware            |             MLU370, MLU590   |
|        2       |Job types                             |          U1                               |
|        3       |Layouts                                |               ARRAY                   |
|        4       |Whether multi-dimensions are supported|            No                      |
|        5       |Whether element zero is supported     |              Yes                        |
|        6       |Data type(half/float)                 |          float           |
|        7       |Whether there is size limit           |                     No                 |

#### 3.1.3 New Feature Test

Since this ops is merged in other repo, we only focus on whether this ops works well.

For all 3 cases of network, this ops passed unit test.

### 3.2 Summary Analysis

For Accurary
3 network case all passed. Already be verified in MMCV network.

For Efficiency
See in MMCV network.

### CODE REVIEW

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/test_case/case_0.prototxt [hunk 1]
```diff
@@ -0,0 +1,280 @@
+op_name: "ms_deform_attn_forward"
+op_type: "MS_DEFORM_ATTN_FORWARD"
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/test_case/case_0.prototxt:2]: 没有op_type

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#ifndef TEST_MLUOP_GTEST_PB_GTEST_SRC_ZOO_MS_DEFORM_ATTN_FORWARD_MS_DEFORM_ATTN_FORWARD_H_   // NOLINT
+#define TEST_MLUOP_GTEST_PB_GTEST_SRC_ZOO_MS_DEFORM_ATTN_FORWARD_MS_DEFORM_ATTN_FORWARD_H_   // NOLINT
+#include <vector>
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.h:25]:
```suggestion

```

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+#include <algorithm>
+#include <cmath>
+#include <string>
+#include <vector>
+#include "math.h"
+
+namespace mluoptest {
+
+float ms_deform_attn_im2col_bilinear(const float *&bottom_data,
+                                     const int &height,
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.cpp:34]: 这个函数改为类内函数是不是好点，如果反向也用到同名函数，是不是有符号冲突

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * - data_value.dims[1] depends on data_spatial_shapes:
+ *   \f$data_value.dims[1]=\sum_{l=1}^{L}H_l*W_l\f$
+ * - data_level_start_index value depends on data_spatial_shapes:
+ *   \f$data_level_start_index[0]=0; data_level_start_index[i]=\sum_{l=1}^{i}H_l*W_l,i=1,...L-1 \f$
+ * - The input \b data_sampling_loc with NaN or infinity is not supported.
+ *
+ * @par Reference
+ * - Deformable DETR: Deformable Transformers for End-to-End Object Detection, Xizhou Zhu, 2020.
+ * - https://arxiv.org/pdf/2010.04159
+ */
```
Comment (tudejiang79) [inline bangc-ops/mlu_op.h:11538]: 这里的注释内容为按照API文档标准写，发给关焕平review下

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+                 handle->nram_size) {
+    return MS_DEFORM_ATTN_FORWARD_DEFAULT;
+  } else if (channels > handle->nram_size / 12 / sizeof(float)) {
+    return MS_DEFORM_ATTN_FORWARD_DEFAULT;
+  } else {
+    return MS_DEFORM_ATTN_FORWARD_SMALL_CHANNEL;
+  }
+}
+
+mluOpStatus_t paramcheck(
```
Comment (tudejiang79) [inline bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu:79]: 这个函数这个命名容易产同名函数冲突

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+    LOG(ERROR) << "[mluOpMsDeformAttnForward] Check failed: element num zero.";
+    return MLUOP_STATUS_BAD_PARAM;
+  }
+  if (num_levels == 0 || num_points == 0) {
+    VLOG(5) << "mluOpFill_v3 start.";
+    const float fill_value = 0.0f;
+    PARAM_CHECK("[mluOpMsDeformAttnForward]",
+        MLUOP_STATUS_SUCCESS ==
+            mluOpFill_v3(handle, MLUOP_POINTER_MODE_HOST, &fill_value,
+                         data_col_desc, data_col));
```
Comment (tudejiang79) [inline bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu:209]: 这里依赖Fill函数，未在independent_build.sh 中说明依赖关系

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
                          const mluOpTensorDescriptor_t grads_image_desc,
                          void *grads_image);
 
+// Group:MsDeformAttnForward
```
Comment (tudejiang79) [inline bangc-ops/mlu_op.h:11477]: 在docs目标下没有设计文档和index.rst的内容

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+/*!
+ * @brief Implements a multi-scale deformable attention module used in Deformable-Detr.
+ * For detailed information about Deformable-Detr, see "Deformable DETR: Deformable
+ * Transformers for End-to-End Object Detection."
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11481]: 引号前的句号写到引号后面

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/test_case/case_0.prototxt (hunk 1 above)
Comment (Unireverse) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/test_case/case_0.prototxt]: done

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.h (hunk 2 above)
Comment (Unireverse) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.h]: done

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.cpp (hunk 3 above)
Comment (Unireverse) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.cpp]: done

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu (hunk 5 above)
Comment (Unireverse) [reply bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu]: done

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu (hunk 6 above)
Comment (Unireverse) [reply bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu:209]: done

File: bangc-ops/mlu_op.h (hunk 7 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h:11477]: done

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues
+ * in ::mluOpMsDeformAttnForward operation. For detailed information, see ::mluOpHandle_t.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11485]: operation > function

Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: bangc-ops/mlu_op.h (hunk 8 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * @param[in] data_value_desc
+ * The descriptor of data_value. For detailed information, see ::mluOpTensorDescriptor_t.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11487]: The descriptor of the tensor \b data_value .

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * @param[in] data_value
+ * Pointer to the MLU memory that stores the input multi-scale feature maps.
+ * @param[in] data_spatial_shapes_desc
+ * The descriptor of data_spatial_shapes. For detailed information, see ::mluOpTensorDescriptor_t.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11491]: The descriptor of the tensor \b data_spatial_shapes .

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * @param[in] data_spatial_shapes
+ * Pointer to the MLU memory that stores the shapes of multi-scale feature maps.
+ * @param[in] data_level_start_index_desc
+ * The descriptor of data_level_start_index. For detailed information,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11495]: The descriptor of the tensor \b data_level_start_index .

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] data_level_start_index
+ * Pointer to the MLU memory that stores the feature maps offset in data_value.
+ * @param[in] data_sampling_loc_desc
+ * The descriptor of data_sampling_loc. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] data_sampling_loc
+ * Pointer to the MLU memory that stores the normalized coordinates of sample points.
+ * @param[in] data_attn_weight_desc
+ * The descriptor of data_attn_weight. For detailed information,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11505]: The descriptor of the tensor \b data_attn_weight .

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] data_attn_weight
+ * Pointer to the MLU memory that stores the attention weight.
+ * @param[in] im2col_step
+ * The value of im2col_step.
+ * @param[in] data_col_desc
+ * The descriptor of the output tensor. For detailed information,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11512]: The descriptor of the tensor \b data_col .

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[out] data_col
+ * Pointer to the MLU memory that stores the output deformable attention feature.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_NOT_SUPPORTED
+ *
+ * @par Data Type
+ * - This function supports the following data types for input and output.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11520]: 上面那句改成The supported data types of input and output tensors are as follows:

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

File: bangc-ops/mlu_op.h (hunk 10 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: bangc-ops/mlu_op.h (hunk 11 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: bangc-ops/mlu_op.h (hunk 12 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * The descriptor of data_spatial_shapes. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] data_spatial_shapes
+ * Pointer to the MLU memory that stores the shapes of multi-scale feature maps.
+ * @param[in] data_level_start_index_desc
+ * The descriptor of data_level_start_index. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] data_level_start_index
+ * Pointer to the MLU memory that stores the feature maps offset in data_value.
+ * @param[in] data_sampling_loc_desc
+ * The descriptor of data_sampling_loc. For detailed information,
```
[the remark this answers is not in the document]
Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: bangc-ops/mlu_op.h (hunk 13 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: bangc-ops/mlu_op.h (hunk 14 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h:11512]: done

File: bangc-ops/mlu_op.h (hunk 15 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (Unireverse) [reply bangc-ops/mlu_op.h:11538]: done

**Review by tudejiang79 — APPROVED** [review]

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md [hunk 17]
```diff
@@ -0,0 +1,483 @@
+# Multi-scale Deformable Attention Forward算子开发设计方案
+* #### 文档基本信息
+| 项目名称    | Training Solution                                            |
+| ----------- | ------------------------------------------------------------ |
+| 算子名称    | ms_deform_attn_forward                                        |
+| JIRA编号    | [CNNLCORE-8312](http://jira.cambricon.com/browse/CNNLCORE-8312)|
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md:6]: 这里

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md [hunk 18]
```diff
@@ -0,0 +1,483 @@
...
+| 输入数据类型| data_value: float32<br>data_spatial_shapes: int32<br>data_level_start_index: int32<br>data_sampling_loc: float32<br>data_attn_weight: float32 |
+| 输入Shape   | data_value: [bs, num_keys, num_heads, channels]<br>data_spatial_shapes: [num_levels, 2]<br>data_level_start_index: [num_levels]<br>data_sampling_loc: [bs, num_queries, num_heads, num_levels, num_points, 2]<br>data_attn_weight: [bs, num_queries, num_heads, num_levels, num_points] |
+| 输入Layout  | ARRAY                                                       |
+| 输出数据类型| data_col: float32                                            |
+| 输出Shape   | data_col: [bs, num_queries, num_heads, channels]            |
+| 输出Layout  | ARRAY                                                        |
+| 是否需要支持原位        | 否                                                 |
+| 是否需要支持stride机制  | 否                                                 |
+| 是否需要支持广播  | 否                                                       |
+| 0元素检查是否直接返回  | 仅在num_levels和num_points维度上支持0元素，此时data_col输出值全0，返回CNNL_STATUS_SUCCESS；其他情况下不支持0元素，返回CNNL_STATUS_BAD_PARAM |
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md:55]: 删除NL字段，其他处相同

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md [hunk 19]
```diff
@@ -0,0 +1,483 @@
+# Multi-scale Deformable Attention Forward算子开发设计方案
+* #### 文档基本信息
+| 项目名称    | Training Solution                                            |
```
Comment (tudejiang79) [inline docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md:3]: 项目名称也不要，按照mlu-ops的设计文档修改

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md [hunk 20]
```diff
@@ -0,0 +1,483 @@
...
+                                       const at::Tensor &sampling_loc,
+                                       const at::Tensor &attn_weight,
+                                       const int im2col_step)
+```
+
+### 2.2 接口设计
+
+```c++
+cnnlStatus_t CNNL_WIN_API cnnlMsDeformAttnForward(cnnlHandle_t handle,
+                                                  const cnnlTensorDescriptor_t data_value_desc,
```
Comment (tudejiang79) [inline docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md:127]: 删除cnnl相关的单词

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md (hunk 17 above)
Comment (Unireverse) [reply docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md]: done

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md (hunk 18 above)
Comment (Unireverse) [reply docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md]: done

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -11474,6 +11474,84 @@ mluOpRoiAlignBackward_v2(mluOpHandle_t handle,
...
+ * @par Data Type
+ * - The supported data types of input and output tensors are as follows:
+ *   - \b data_value: float
+ *   - \b data_spatial_shapes: int32
+ *   - \b data_level_start_index: int32
+ *   - \b data_sampling_loc: float
+ *   - \b data_attn_weight: float
+ *   - \b data_col: float
+ *
+ * @note
```
[the remark this answers is not in the document]
Comment (Unireverse) [reply bangc-ops/mlu_op.h]: done

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md (hunk 19 above)
Comment (Unireverse) [reply docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md]: done

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md (hunk 20 above)
Comment (Unireverse) [reply docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md]: done

**Review by WangYuan-cn — APPROVED** [review]

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.cpp [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+      }
+    }
+    *data_col_ptr = col;
+  }
+  return;
+}
+
+void MsDeformAttnForwardExecutor::paramCheck() {
+  GTEST_CHECK(parser_->getInputNum() == 5);
+  GTEST_CHECK(parser_->getOutputNum() == 1);
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_forward/ms_deform_attn_forward.cpp:142]: #define GTEST_CHECK(condition, ...)                                 \
  if (unlikely(!(condition))) {                                     \
    ADD_FAILURE() << "Check failed: " #condition ". " #__VA_ARGS__; \
    throw std::invalid_argument(std::string(__FILE__) + " +" +      \
                                std::to_string(__LINE__));          \
  }

这样看GTEST_CHECK(condition，这里加个一个字符串说明为什么错误是不是好点)

**Review by tudejiang79 — APPROVED** [review]

**Review by GraceGuanH — APPROVED** [review]

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 23]
```diff
@@ -719,3 +719,8 @@ mluOpPointsInBoxes
 mluOpRoiAlignBackward
 ---------------------------------
 该算子是RoiAlignForward算子的反向，根据 boxes中的坐标值，使用 spatial_scale 参数进行缩放，计算出 Roi窗口的坐标、长宽。pool_mode等于0时，为Max模式的反向，按照argmax_x 和 argmax_y 的坐标，进行双线性插值，计算映射到 grad_image 上坐标点的权重，分别对grad_output加权后，累加反传梯度；pool_mode等于1时，为Avg模式的反向，根据 sampling_ratio 参数，计算每个 grad_output 需要反传梯度的采样点数，再计算每个采样点的x, y坐标，进行双线性插值，对grad_output加权、均摊，累加反传梯度。
+
```
Comment (wushaoqiang2168) [inline docs/bangc-docs/user_guide/9_operators/index.rst:722]: 补充：.. _mluOpMsDeformAttnForward:

**Review by DanieeelLiu — APPROVED** [review]

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md [hunk 24]
```diff
@@ -0,0 +1,481 @@
...
+本文档为`ms_deform_attn_forward`算子的设计文档，包括需求分析、接口设计、方案设计、性能优化记录和方案实施部分。
+* #### 算子需求checklist
+算子需求提出者需要`提供`的信息如下：
+- 框架负责人
+- 算子接口描述
+- 功能描述
+- 框架版本 + 对应源码路径
+- 需求对应网络
+- 网络中用到的规模
+- 常用规模下的竞品性能（可选）
```
Comment (DanieeelLiu) [inline docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md:23]: 竞品性能删除

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md [hunk 25]
```diff
@@ -0,0 +1,481 @@
...
+| 数据类型限制| data_value, data_sampling_loc, data_attn_weight, data_col 所指向元素的数据类型为float，data_spatial_shapes, data_level_start_index 所指向元素的数据类型为int32 |
+| 数据范围限制| $`data\_value.dims[1]=\sum_{l=1}^{L}H_l*W_l`$<br>data_level_start_index由data_spatial_shapes计算得到，$`data\_level\_start\_index[0]=0; data\_level\_start\_index[i]=\sum_{l=1}^{i}H_l*W_l,i=1,...L-1`$ |
+| 原位限制    | 不支持原位       |
+| stride限制  | 不支持stride机制 |
+| 广播限制    | 不支持广播        |
+### 1.5 验收标准
+#### 1.5.1 精度验收标准
+- MLUOP精度验收标准：输出data_col适配动态阈值，采用动态阈值标准：diffs = [diff1, diff2, diff4]，threshold_rate = [10, 10, 1]。
+#### 1.5.2 性能验收标准
+- 竞品性能分析：
```
Comment (DanieeelLiu) [inline docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md:97]: 同上，替换成 参考接口

File: docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md [hunk 26]
```diff
@@ -0,0 +1,481 @@
...
+| Tesla V100-SXM2 16GB | mmcv-1.5.1 + cuda-10.2 + pytorch-1.9 + driver-440.64.00 | data_value: float32<br>data_spatial_shapes: int32<br>data_level_start_index: int32<br>data_sampling_loc: float32<br>data_attn_weight: float32<br>data_col: float32 | data_value: [1, 40000, 8, 32]<br>data_spatial_shapes: [1, 2]<br>data_level_start_index: [1]<br>data_sampling_loc: [1, 900, 8, 1, 4, 2]<br>data_attn_weight: [1, 900, 8, 1, 4]<br>data_col: [1, 900, 8, 32]  | 28.257790% | 67.967224% | 29.856us |
+| Tesla V100-SXM2 16GB | mmcv-1.5.1 + cuda-10.2 + pytorch-1.9 + driver-440.64.00 | data_value: float32<br>data_spatial_shapes: int32<br>data_level_start_index: int32<br>data_sampling_loc: float32<br>data_attn_weight: float32<br>data_col: float32 | data_value: [6, 30825, 8, 32]<br>data_spatial_shapes: [4, 2]<br>data_level_start_index: [4]<br>data_sampling_loc: [6, 9664, 8, 4, 8, 2]<br>data_attn_weight: [6, 9664, 8, 4, 8]<br>data_col: [6, 9664, 8, 32]  | 34.306033% | 82.415587% | 2.467792ms |- 网络中使用到的规模性能优于或至少与竞品性能持平。
+- 部分与竞品差距过大的规模在4.算子性能优化记录中进行说明。
```
Comment (DanieeelLiu) [inline docs/bangc-docs/design_docs/ms_deform_attn_forward/ms_deform_attn_forward.md:104]: 同上

**Pull request merged**